packages/next/client/route-loader.ts

```diff
@@ -164,13 +164,15 @@ function resolvePromiseWithTimeout<T>(
       resolve(r)
     }).catch(reject)
 
-    requestIdleCallback(() =>
-      setTimeout(() => {
-        if (!cancelled) {
-          reject(err)
-        }
-      }, ms)
-    )
+    if (process.env.NODE_ENV !== 'development') {
```

Comment:
Instead of not timing out in development when a page is being compiled we could check the [HMR events](https://github.com/vercel/next.js/blob/402f0360cece821b9e0017f6ecb499d129993600/packages/next/client/dev/error-overlay/hot-dev-client.js#L203-L208) and delay the timeout until after the compilation is complete. 

This should allow the full-page transition timeout to trigger correctly in development when data-fetching or other requests take a while so that you don't have to wait until a production build to notice the timeout is being hit on transition.